Documentation updates

teobiton opened the pull request:
This PR addresses small issues of the documentation:

- #3940
- #3694 
- #2981

<!--

Thanks for improving cocotb! Here are some points to make this as smooth as possible.
Not all of them may be applicable.

Most important: please explain *why* you are proposing this change.

* Make sure you have read https://github.com/cocotb/cocotb/blob/master/CONTRIBUTING.md
* Extend or add a test under `tests/test_cases/`.
* Add documentation under `docs/source/`,
  docstrings in Python code, or Doxygen markup in C/C++ code.
  Use ``versionadded``/``versionchanged``/``deprecated``.
* Add a newsfragment - see `docs/source/newsfragments/README.rst`.
* Use `closes #XXXX` to auto-close the issue that this PR fixes (if such).

-->

cmarqu commented on docs/source/conf.py:
```diff
@@ -84,6 +84,10 @@
 v_major, v_minor = LooseVersion(release).version[:2]
 version = "{}.{}".format(v_major, v_minor)
+# Cocotb commit ID
+commit_id = (
```
I think it would be good to also (still) support building the documentation from a tarball without git installed.

cmarqu commented on docs/source/building.rst:
```diff
@@ -54,11 +54,19 @@ Cocotb
 
     .. versionchanged:: 1.6.0 Strip leading and trailing whitespace
 
+    .. deprecated:: 2.0
+
+        :envvar:`TOPLEVEL` is renamed to :envvar:`COCOTB_TOPLEVEL`.
```
You aren't actually introducing the new names to the list of variables yet, are you? (Or is the GH client on my cell phone fooling me?)

teobiton replied:
No you're right I just added these messages for each that was renamed, I don't know if that's enough though. My other idea was to introduce the new names in the list and make the old ones point to their description instead.

cmarqu replied:
I also think that using the new names and making the old ones be pointers is the best solution.

ktbarrett commented on docs/source/writing_testbenches.rst:
```diff
@@ -230,6 +230,15 @@ the various actions described in :ref:`assignment-methods` can be used.
+
+    Some actions do not work on all simulators:
+
+    - Release does not work on GHDL (see :issue:`3830`).
+    - Force and Release do not work on Verilator (see :issue:`3831`).
+    - Riviera's VPI implicitly releases signal when overwriting forced signal with normal deposit (see :issue:`3832`).
+    - Questa's FLI allows overwriting forced signal with normal deposit (see :issue:`3833`).
```
I like the thoroughness, but we have to remember to delete these lines when upstream fixes the issue. We either need to make this more general like "not all simulators fully support this feature, see the issues page" and link a URL that does a search on the issues that returns these PRs. ***OR*** we can just add comments to those issues to update the documentation so we don't forget.

teobiton replied:
I'd go with the second option, that way it's explicitly written in the documentation and then the lines can be deleted when upstream issues are resolved. I can add comments on the issues once this is merged.

cmarqu replied:
It may very well be the case that some simulators fix their behavior in newer versions even for older cocotb versions, and then our  published docs would not be accurate anymore...
We have links to issues based on labels for each supported simulator in the Simulator Support section.
Maybe it's enough to mention those links here?

teobiton commented:
Just updated the documentation and added the list of new variables. It feels weird to have `COCOTB_TESTCASES` as a new variable but it's also deprecated at the same time.
Plus, if I change to the new names everywhere in the documentation, I also have to change it in the quick-start examples and such, is that something we want now or later ?

ktbarrett commented:
@teobiton Yeah you'll need to update all the existing refs. That's expected. As for the `.. deprecated::` label, I might instead use `.. versionchanged:: 2.0`. What say you @cmarqu?

cmarqu commented:
> @teobiton Yeah you'll need to update all the existing refs. That's expected. As for the `.. deprecated::` label, I might instead use `.. versionchanged:: 2.0`. What say you @cmarqu?

The description of https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated sounds fitting here IMO.

teobiton commented:
Just updated all references to the old variables that appear in the documentation and left everything else as is.

cmarqu commented on docs/source/custom_flows.rst:
```diff
@@ -16,7 +16,7 @@ this chapter shows the minimum settings to be done.
 
 * Define :envvar:`LIBPYTHON_LOC` using ``$(cocotb-config --libpython)``.
-* Define :envvar:`MODULE` with the name of the Python module containing your testcases.
+* Define :envvar:`COCOTB_TEST_MODULES` with the name of the Python module containing your testcases.
```
```suggestion
* Define :envvar:`COCOTB_TEST_MODULES` with the name of the Python module(s) containing your testcases.
```

cmarqu commented on docs/source/quickstart.rst:
```suggestion
* and Python modules that contain our cocotb tests (:envvar:`COCOTB_TEST_MODULES`).
```

cmarqu commented on docs/source/writing_testbenches.rst:
```diff
@@ -230,6 +230,10 @@ the various actions described in :ref:`assignment-methods` can be used.
 
+.. warning::
+
+    Not all simulators support these features; refer to the :ref:`simulator-support` section for details or to `issues with label upstream <https://github.com/cocotb/cocotb/issues?q=is%3Aissue+-label%3Astatus%3Aduplicate+label%3Aupstream>`_
```
```suggestion
    Not all simulators support these features; refer to the :ref:`simulator-support` section for details or to `issues with label "upstream" <https://github.com/cocotb/cocotb/issues?q=is%3Aissue+-label%3Astatus%3Aduplicate+label%3Aupstream>`_
```

cmarqu commented on docs/source/building.rst:
```diff
@@ -225,8 +278,18 @@ Regression Manager
 
+    .. deprecated:: 2.0
+
+        :envvar:`COVERAGE_RCFILE` is renamed to :envvar:`COCOTB_COVERAGE_RCFILE`.
```
While this PR is documenting the current code correctly, I think the PR that did the renaming of this variable in the code has slightly broken the coverage functionality.
Unless I'm misunderstanding things, we did rely on `coverage.py` to read the original name `COVERAGE_RCFILE` (see https://coverage.readthedocs.io/en/latest/config.html#configuration-reference), which it won't do for `COCOTB_COVERAGE_RCFILE`.

cmarqu replied:
I opened #3947 for this.

ktbarrett commented on docs/source/building.rst:
```diff
@@ -154,22 +174,36 @@ Cocotb
+
+        :envvar:`MODULE` is renamed to :envvar:`COCOTB_TEST_MODULES`.
+
+.. envvar:: COCOTB_TESTCASE
 
     A comma-separated list of tests to run.
     Does an exact match on the test name.
 
+    .. versionadded:: 2.0
+
     .. deprecated:: 2.0
 
+        :envvar:`COCOTB_TESTCASE` is renamed to :envvar:`COCOTB_TESTCASE`.
+
```
This doesn't match up with the others.

teobiton replied:
Oops, fixed it.

ktbarrett reviewed (commented):
I don't think we should keep the old envvars around; they aren't intended to be a part of the API anymore. I think just a deprecated message mentioning the old variables name is sufficient.

teobiton commented:
> I don't think we should keep the old envvars around; they aren't intended to be a part of the API anymore. I think just a deprecated message mentioning the old variables name is sufficient.

Okay but then where does the deprecation message go ? With the new variables ?

ktbarrett commented:
> With the new variables ?

Yeah. It's a feature related to the thing deprecated. Not all deprecated blcoks are deprecating the whole feature. Alternatively I like versionchanged. Or both? One versionchanged for the rename, another deprecation mentioning the old name is now a deprecated alias and will be removed.
